### PR TITLE
[SPARK-39195][SQL] Spark OutputCommitCoordinator should abort stage when committed file not consistent with task status

### PR DESCRIPTION
**Pull request by AngersZhuuuu**
### What changes were proposed in this pull request?
There is a case for current code may cause data correctness issue: 
  1. One writing data task commit task success, but the container was killed after commit task, so this task failed, but the data file still remained under committedTaskPath .
  2. DAGScheduler call `handleCompletedTask` and then call `outputCommitCoordinator.taskCompleted`,  outputCommitCoordinator will remove the lock of failed task's partition.
  3. Then failed task's rerun with an new attempt,  the new attempt task call `outputCommitCoordinator.canCommit()` will return true since the lock of this partition had been removed, then it commit task success, also task final  succeed.
  4. Two files remained under this job's attempt path for  same partition.
  5. CommitJob commit both two committed task path's data.
  6. Finally data duplicated.

In this pr, we do below since:

1. When commit task success, executor send an CommitOutputSuccess message to  outputCommitCoordinator.
2. When outputCommitCoordinator handle `taskComplete`, if task failed but commit success, means data duplicate will happen,  we should failed to job.

### Why are the changes needed?
Fix data duplicated issue.

### Does this PR introduce _any_ user-facing change?
No

### CODE REVIEW

**Comment by AngersZhuuuu** [comment]
gentle ping @dongjoon-hyun @cloud-fan @HyukjinKwon @srowen Could you take a view of this data correctness issue?

File: core/src/main/scala/org/apache/spark/scheduler/OutputCommitCoordinator.scala
Comment (AngersZhuuuu) [inline core/src/main/scala/org/apache/spark/scheduler/OutputCommitCoordinator.scala:259]: Seems throw exception here won't stop the job, any suggestion.

**Comment by AngersZhuuuu** [comment]
ping @cloud-fan Could you take a look?

File: core/src/main/scala/org/apache/spark/scheduler/DAGScheduler.scala
Comment (cloud-fan) [reply core/src/main/scala/org/apache/spark/scheduler/DAGScheduler.scala:2596]: Actually, we can generalize `taskSetFailed` a little bit, as what it does is just aborting a stage.

File: core/src/main/scala/org/apache/spark/SparkContext.scala
Comment (cloud-fan) [inline core/src/main/scala/org/apache/spark/SparkContext.scala:471]: this looks weird, I think we just need to pass `this` to `createSparkEnv`, which creates driver-side `SparkEnv` that contains `OutputCommitCoordinator`

Comment (AngersZhuuuu) [reply core/src/main/scala/org/apache/spark/SparkContext.scala]: > this looks weird, I think we just need to pass `this` to `createSparkEnv`, which creates driver-side `SparkEnv` that contains `OutputCommitCoordinator`

How about current?

File: core/src/main/scala/org/apache/spark/SparkEnv.scala
Comment (cloud-fan) [inline core/src/main/scala/org/apache/spark/SparkEnv.scala:435]: this can be put right after the parameter list

File: core/src/test/scala/org/apache/spark/scheduler/OutputCommitCoordinatorSuite.scala
Comment (cloud-fan) [inline core/src/test/scala/org/apache/spark/scheduler/OutputCommitCoordinatorSuite.scala:191]:
```suggestion
    // A new task should not be allowed to become stage failed because of potential data duplication
```

File: core/src/test/scala/org/apache/spark/scheduler/OutputCommitCoordinatorSuite.scala
Comment (cloud-fan) [inline core/src/test/scala/org/apache/spark/scheduler/OutputCommitCoordinatorSuite.scala:239]: ditto

File: core/src/test/scala/org/apache/spark/scheduler/OutputCommitCoordinatorSuite.scala
Comment (cloud-fan) [inline core/src/test/scala/org/apache/spark/scheduler/OutputCommitCoordinatorSuite.scala:191]: We need this level of this for the new behavior

File: core/src/main/scala/org/apache/spark/SparkContext.scala
Comment (AngersZhuuuu) [reply core/src/main/scala/org/apache/spark/SparkContext.scala]: Done

File: core/src/test/scala/org/apache/spark/scheduler/OutputCommitCoordinatorSuite.scala
Comment (AngersZhuuuu) [reply core/src/test/scala/org/apache/spark/scheduler/OutputCommitCoordinatorSuite.scala:191]: > We need this level of this for the new behavior

How about current?

**Comment by AngersZhuuuu** [comment]
Doc failed seems not related to this pr, any more suggestion?

File: core/src/test/scala/org/apache/spark/scheduler/OutputCommitCoordinatorIntegrationSuite.scala
Comment (cloud-fan) [inline core/src/test/scala/org/apache/spark/scheduler/OutputCommitCoordinatorIntegrationSuite.scala:49]: we should still keep this comment

File: core/src/test/scala/org/apache/spark/scheduler/OutputCommitCoordinatorIntegrationSuite.scala
Comment (cloud-fan) [inline core/src/test/scala/org/apache/spark/scheduler/OutputCommitCoordinatorIntegrationSuite.scala:60]: why this change? I think the original test wanted to test the case that we failed before committing in the first task attempt.

File: core/src/test/scala/org/apache/spark/scheduler/OutputCommitCoordinatorSuite.scala
Comment (cloud-fan) [inline core/src/test/scala/org/apache/spark/scheduler/OutputCommitCoordinatorSuite.scala:238]: we need to update the code comment before this code chunk.

File: core/src/test/scala/org/apache/spark/scheduler/OutputCommitCoordinatorSuite.scala
Comment (cloud-fan) [inline core/src/test/scala/org/apache/spark/scheduler/OutputCommitCoordinatorSuite.scala:292]: I'm a bit worried about changing the behavior for the existing testing output committer. Can we create a new one for the new test case?

File: core/src/test/scala/org/apache/spark/scheduler/OutputCommitCoordinatorIntegrationSuite.scala
Comment (AngersZhuuuu) [reply core/src/test/scala/org/apache/spark/scheduler/OutputCommitCoordinatorIntegrationSuite.scala:49]: Add back

File: core/src/test/scala/org/apache/spark/scheduler/OutputCommitCoordinatorSuite.scala
Comment (AngersZhuuuu) [reply core/src/test/scala/org/apache/spark/scheduler/OutputCommitCoordinatorSuite.scala:292]: Revert this change, not need this change.

File: core/src/test/scala/org/apache/spark/scheduler/OutputCommitCoordinatorIntegrationSuite.scala
Comment (AngersZhuuuu) [reply core/src/test/scala/org/apache/spark/scheduler/OutputCommitCoordinatorIntegrationSuite.scala:60]: revert now

File: core/src/test/scala/org/apache/spark/scheduler/OutputCommitCoordinatorSuite.scala
Comment (AngersZhuuuu) [reply core/src/test/scala/org/apache/spark/scheduler/OutputCommitCoordinatorSuite.scala:238]: How about current?

**Review by cloud-fan — APPROVED** [review]
LGTM if tests pass

File: core/src/test/scala/org/apache/spark/scheduler/OutputCommitCoordinatorSuite.scala
Comment (cloud-fan) [inline core/src/test/scala/org/apache/spark/scheduler/OutputCommitCoordinatorSuite.scala:267]: is this test the same as https://github.com/apache/spark/pull/36564/files#diff-d5d8811dcf8af13683cb2bac4a583577cb2accca0cd64182cf214bbd174b017cR47 ?

Comment (AngersZhuuuu) [reply core/src/test/scala/org/apache/spark/scheduler/OutputCommitCoordinatorSuite.scala]: Removed

**Comment by AngersZhuuuu** [comment]
> LGTM if tests pass

GA failed not related to this PR

```
__w/spark/spark/docs/_plugins/copy_api_dirs.rb:130:in `<top (required)>': Python doc generation failed (RuntimeError)
2022-06-11T07:16:53.8252035Z 	from /__w/spark/spark/docs/.local_ruby_bundle/ruby/2.7.0/gems/jekyll-4.2.1/lib/jekyll/external.rb:60:in `require'
2022-06-11T07:16:53.8252686Z 	from /__w/spark/spark/docs/.local_ruby_bundle/ruby/2.7.0/gems/jekyll-4.2.1/lib/jekyll/external.rb:60:in `block in require_with_graceful_fail'
2022-06-11T07:16:53.8253306Z 	from /__w/spark/spark/docs/.local_ruby_bundle/ruby/2.7.0/gems/jekyll-4.2.1/lib/jekyll/external.rb:57:in `each'
2022-06-11T07:16:53.8253947Z 	from /__w/spark/spark/docs/.local_ruby_bundle/ruby/2.7.0/gems/jekyll-4.2.1/lib/jekyll/external.rb:57:in `require_with_graceful_fail'
2022-06-11T07:16:53.8254613Z 	from /__w/spark/spark/docs/.local_ruby_bundle/ruby/2.7.0/gems/jekyll-4.2.1/lib/jekyll/plugin_manager.rb:89:in `block in require_plugin_files'
2022-06-11T07:16:53.8255235Z 	from /__w/spark/spark/docs/.local_ruby_bundle/ruby/2.7.0/gems/jekyll-4.2.1/lib/jekyll/plugin_manager.rb:87:in `each'
2022-06-11T07:16:53.8255870Z 	from /__w/spark/spark/docs/.local_ruby_bundle/ruby/2.7.0/gems/jekyll-4.2.1/lib/jekyll/plugin_manager.rb:87:in `require_plugin_files'
2022-06-11T07:16:53.8256520Z 	from /__w/spark/spark/docs/.local_ruby_bundle/ruby/2.7.0/gems/jekyll-4.2.1/lib/jekyll/plugin_manager.rb:21:in `conscientious_require'
2022-06-11T07:16:53.8257116Z 	from /__w/spark/spark/docs/.local_ruby_bundle/ruby/2.7.0/gems/jekyll-4.2.1/lib/jekyll/site.rb:131:in `setup'
2022-06-11T07:16:53.8257695Z 	from /__w/spark/spark/docs/.local_ruby_bundle/ruby/2.7.0/gems/jekyll-4.2.1/lib/jekyll/site.rb:36:in `initialize'
2022-06-11T07:16:53.8258278Z 	from /__w/spark/spark/docs/.local_ruby_bundle/ruby/2.7.0/gems/jekyll-4.2.1/lib/jekyll/commands/build.rb:30:in `new'
2022-06-11T07:16:53.8258888Z 	from /__w/spark/spark/docs/.local_ruby_bundle/ruby/2.7.0/gems/jekyll-4.2.1/lib/jekyll/commands/build.rb:30:in `process'
2022-06-11T07:16:53.8259731Z 	from /__w/spark/spark/docs/.local_ruby_bundle/ruby/2.7.0/gems/jekyll-4.2.1/lib/jekyll/command.rb:91:in `block in process_with_graceful_fail'
2022-06-11T07:16:53.8260335Z 	from /__w/spark/spark/docs/.local_ruby_bundle/ruby/2.7.0/gems/jekyll-4.2.1/lib/jekyll/command.rb:91:in `each'
2022-06-11T07:16:53.8260952Z 	from /__w/spark/spark/docs/.local_ruby_bundle/ruby/2.7.0/gems/jekyll-4.2.1/lib/jekyll/command.rb:91:in `process_with_graceful_fail'
2022-06-11T07:16:53.8261625Z 	from /__w/spark/spark/docs/.local_ruby_bundle/ruby/2.7.0/gems/jekyll-4.2.1/lib/jekyll/commands/build.rb:18:in `block (2 levels) in init_with_program'
2022-06-11T07:16:53.8262274Z 	from /__w/spark/spark/docs/.local_ruby_bundle/ruby/2.7.0/gems/mercenary-0.4.0/lib/mercenary/command.rb:221:in `block in execute'
2022-06-11T07:16:53.8262921Z 	from /__w/spark/spark/docs/.local_ruby_bundle/ruby/2.7.0/gems/mercenary-0.4.0/lib/mercenary/command.rb:221:in `each'
2022-06-11T07:16:53.8263524Z 	from /__w/spark/spark/docs/.local_ruby_bundle/ruby/2.7.0/gems/mercenary-0.4.0/lib/mercenary/command.rb:221:in `execute'
2022-06-11T07:16:53.8264134Z 	from /__w/spark/spark/docs/.local_ruby_bundle/ruby/2.7.0/gems/mercenary-0.4.0/lib/mercenary/program.rb:44:in `go'
2022-06-11T07:16:53.8264718Z 	from /__w/spark/spark/docs/.local_ruby_bundle/ruby/2.7.0/gems/mercenary-0.4.0/lib/mercenary.rb:21:in `program'
2022-06-11T07:16:53.8265284Z 	from /__w/spark/spark/docs/.local_ruby_bundle/ruby/2.7.0/gems/jekyll-4.2.1/exe/jekyll:15:in `<top (required)>'
2022-06-11T07:16:53.8265698Z 	from /__w/spark/spark/docs/.local_ruby_bundle/ruby/2.7.0/bin/jekyll:23:in `load'
2022-06-11T07:16:53.8266159Z 	from /__w/spark/spark/docs/.local_ruby_bundle/ruby/2.7.0/bin/jekyll:23:in `<main>'
```

File: core/src/test/scala/org/apache/spark/scheduler/OutputCommitCoordinatorIntegrationSuite.scala
Comment (cloud-fan) [inline core/src/test/scala/org/apache/spark/scheduler/OutputCommitCoordinatorIntegrationSuite.scala:47]: since we changed this test, let's put the JIRA ID in the test name

Comment (AngersZhuuuu) [reply core/src/test/scala/org/apache/spark/scheduler/OutputCommitCoordinatorIntegrationSuite.scala]: > since we changed this test, let's put the JIRA ID in the test name

Done

File: core/src/test/scala/org/apache/spark/scheduler/OutputCommitCoordinatorSuite.scala
Comment (AngersZhuuuu) [inline core/src/test/scala/org/apache/spark/scheduler/OutputCommitCoordinatorSuite.scala:146]: @cloud-fan This behavior changed too, so maybe we should remove this?

Comment (cloud-fan) [reply core/src/test/scala/org/apache/spark/scheduler/OutputCommitCoordinatorSuite.scala:146]: SGTM

File: core/src/test/scala/org/apache/spark/scheduler/OutputCommitCoordinatorSuite.scala
Comment (cloud-fan) [reply core/src/test/scala/org/apache/spark/scheduler/OutputCommitCoordinatorSuite.scala:238]: It has not been updated yet
```
    // Commit the 1st attempt, fail the 2nd attempt, make sure 3rd attempt cannot commit,
    // then fail the 1st attempt and make sure the 4th one can commit again.
```

Comment (AngersZhuuuu) [reply core/src/test/scala/org/apache/spark/scheduler/OutputCommitCoordinatorSuite.scala:238]: Updated

**Review by cloud-fan — APPROVED** [review]

**Comment by AngersZhuuuu** [comment]
ping @cloud-fan

**Comment by HyukjinKwon** [comment]
I see the same test failed above https://github.com/AngersZhuuuu/spark/runs/6893999574?check_suite_focus=true#step:9:15747

**Comment by AngersZhuuuu** [comment]
> @AngersZhuuuu Mind taking a look please?

Sure

**Comment by HyukjinKwon** [comment]
@AngersZhuuuu mind taking a look https://issues.apache.org/jira/browse/SPARK-39622? I don't think there's any other committer release changes except this .. cc @HeartSaVioR FYI

**Comment by AngersZhuuuu** [comment]
Yea

**Comment by mstebelev** [comment]
Hi @AngersZhuuuu. I came across problems with this changes after updating to spark 3.4
I write data to iceberg table with S3 backend and the data upload happens in dataWriter.commit() after coordinator.canCommit was called. So if uploading to S3 fails for some reason, the task fails and the partition data can't be uploaded anymore event in the task's retries, because the failed task remains to be the assigned commiter.
Looks like usually data writing is idempotent, because each partition is written into a separate file and you always can do it again in retrying task without data duplication.

**Comment by cloud-fan** [comment]
I'm surprised that iceberg does not overwrite https://github.com/apache/spark/blob/master/sql/catalyst/src/main/java/org/apache/spark/sql/connector/write/BatchWrite.java#L63

cc @huaxingao @dongjoon-hyun

**Comment by mstebelev** [comment]
🤔 so the option useCommitCoordinator is for appending to the same destination, like inserting rows to a remote database and we don't need it in iceberg?

**Comment by huaxingao** [comment]
@cloud-fan Thanks for pinging me. It appears that Iceberg doesn't override this `useCommitCoordinator`. I will take a look at this issue.

**Comment by aokolnychyi** [comment]
@huaxingao @cloud-fan, could you confirm only a single `WriterCommitMessage` will be passed in case of speculative execution even without the commit coordinator? Based on what I see in `V2TableWriteExec`, if multiple task attempts succeed, the last one will win by replacing the commit message from the first one. Am I right?

**Comment by cloud-fan** [comment]
@aokolnychyi In most cases, yes. However, we have `BatchWrite#onDataWriterCommit`, so implementations can decide how to deal with conflicting commit messages, maybe first-win.

**Comment by aokolnychyi** [comment]
Thanks for confirming, @cloud-fan!

File: core/src/main/scala/org/apache/spark/scheduler/OutputCommitCoordinator.scala
Comment (viirya) [inline core/src/main/scala/org/apache/spark/scheduler/OutputCommitCoordinator.scala:164]: @cloud-fan I think this is not very clear or correct in the reason string. `stageState.authorizedCommitters` records a commit is allowed but it is not actually successful. So as you said the driver never knows if the task commit is successful or not. Maybe we should update this to reduce confusion.

Comment (cloud-fan) [reply core/src/main/scala/org/apache/spark/scheduler/OutputCommitCoordinator.scala:164]: +1. @AngersZhuuuu can you refine it?

**Comment by dongjoon-hyun** [comment]
To @AngersZhuuuu , @viirya , @cloud-fan . 
Since this PR is too old, shall we file a JIRA issue for that comment, https://github.com/apache/spark/pull/36564#discussion_r1598660630 ?

**Comment by viirya** [comment]
> To @AngersZhuuuu , @viirya , @cloud-fan . Since this PR is too old, shall we file a JIRA issue for that comment, [#36564 (comment)](https://github.com/apache/spark/pull/36564#discussion_r1598660630) ?

Yea, I think so.

**Comment by viirya** [comment]
I created https://issues.apache.org/jira/browse/SPARK-48292

File: core/src/main/scala/org/apache/spark/scheduler/OutputCommitCoordinator.scala
Comment (AngersZhuuuu) [reply core/src/main/scala/org/apache/spark/scheduler/OutputCommitCoordinator.scala:164]: With https://github.com/apache/spark/pull/38980 seems we didn't need this patch anymore

Comment (AngersZhuuuu) [reply core/src/main/scala/org/apache/spark/scheduler/OutputCommitCoordinator.scala:164]: > oh, shall we create a PR to revert it then?

Double checked with @boneanxs , we should revert this, let me do this.

**Comment by steveloughran** [comment]
Task commits MUST be atomic and a second attempt MUST be able to supercede the first. 

Hadoop FileOutputCommitter v1 (on hdfs, localfs, abfs, *not* gcs) uses atomic rename for this, deleting task completed data path first. A retry will delete that dest path and rename their own work directory to it.

S3A committers PUT a manifest to the task path; relies on PUT being atomic.
Mapreduce Manifest committer does write of manifest and then rename; this is atomic on GCS so safe there (dir rename is non atomic)

The application MUST NOT constrain which of two attempts told to commit succeeds, only that the second one MUST report success.

Why so? Because if task attempt TA1 stops responding after being told to commit, then TA2 will be told to commit, and reports success. But if TA1 is somehow suspended and performs its atomic commit after TA2, then it will be the TA1 manifest which is processed.

If you are encountering problems with jobs where task failures are unrecoverable, that means there is something wrong with the task commit algorithm.

What were you seeing it with?

**Comment by dongjoon-hyun** [comment]
To @steveloughran , for the better context, you had better see and post on https://github.com/apache/spark/pull/46696 instead of here.

BTW, for me, I can say that I don't see any issue with both SPARK-39195 and its revert (SPARK-48292) with S3 environment.

**Comment by akki** [comment]
Hi all

I am facing [this issue](https://stackoverflow.com/q/78789779/3061686) after upgrading to Spark3.5.1 and wonder if this change is a root cause for it. Can anyone here confirm?

Many of our jobs are failing with "Authorized committer" errors and we might have to revert our whole system back to Spark3.3, which would be a lot of work. So I am trying to understand the root issue a bit more before I have to go back reverting my whole Spark+Hadoop system.

Thanks!

**Comment by dfit99** [comment]
Hi, I have been facing a similar issue as @akki  after migrating to Spark 3.5.0

**Pull request closed without merging**
